drivers/adin2111/bm_adin2111.c

```diff
@@ -0,0 +1,272 @@
+    err = BmENOMEM;
+    goto end;
+  }
+  memset(buffer_description, 0, sizeof(adi_eth_BufDesc_t));
+  buffer_description->pBuf = aligned_malloc(DMA_ALIGN_SIZE, length);
+  if (!buffer_description->pBuf) {
+    bm_free(buffer_description);
+    err = BmENOMEM;
+    goto end;
+  }
+  memcpy(buffer_description->pBuf, data, length);
+  buffer_description->bufSize = length;
+  buffer_description->cbFunc = tx_complete;
+  adin2111_SubmitTxBuffer(self->device_handle, ADIN2111_TX_PORT_FLOOD,
+                          buffer_description);
+end:
+  return err;
+}
+
+// Trait wrapper function to convert self from void* to Adin2111*
+static inline BmErr adin2111_netif_send_(void *self, uint8_t *data,
+                                         size_t length) {
+  return adin2111_netif_send(self, data, length);
+}
+
+// Called by the driver on received data
+// If the user has registered a callback, call it
+static void receive_callback_(void *device_param, uint32_t event,
+                              void *buffer_description_param) {
+  (void)device_param;
+  (void)event;
+
+  if (ADIN2111 && ADIN2111->callbacks && ADIN2111->callbacks->receive) {
+    adi_eth_BufDesc_t *buffer_description =
+        (adi_eth_BufDesc_t *)buffer_description_param;
+    uint8_t port_index = 1 << buffer_description->port;
+    ADIN2111->callbacks->receive(port_index, buffer_description->pBuf,
+                                 buffer_description->bufSize);
+  }
+}
+
+/**************** Public API Functions ****************/
+
+/*! @brief Initialize an Adin2111 device
+    @param self a pointer to an Adin2111 struct
+    @return BmOK if successful, otherwise an error */
+BmErr adin2111_init(Adin2111 *self) {
+  BmErr err = BmOK;
+
+  if (!self) {
+    err = BmEINVAL;
+    goto end;
+  }
+
+  if (ADIN2111) {
+    err = BmEALREADY;
+    goto end;
+  }
+
+  ADIN2111 = self;
+  self->device_handle = &DEVICE_STRUCT;
+
+  for (int i = 0; i < RX_QUEUE_NUM_ENTRIES; i++) {
+    adi_eth_BufDesc_t *buffer_description = &RX_BUFFERS[i];
+    memset(buffer_description, 0, sizeof(adi_eth_BufDesc_t));
+    buffer_description->pBuf =
+        aligned_malloc(DMA_ALIGN_SIZE, MAX_FRAME_BUF_SIZE);
+    if (!buffer_description->pBuf) {
+      err = BmENOMEM;
+      goto end;
+    }
+    memset(buffer_description->pBuf, 0, MAX_FRAME_BUF_SIZE);
+    buffer_description->bufSize = MAX_FRAME_BUF_SIZE;
+    buffer_description->cbFunc = receive_callback_;
+  }
+
+  err = adin2111_netif_enable_(self);
+
+end:
+  return err;
+}
+
+/// Build a generic NetworkInterface out of a concrete Adin2111
+NetworkInterface prep_adin2111_netif(Adin2111 *self) {
+  // Create the vtable once and attach a pointer to it every time
+  static NetworkInterfaceTrait const trait = {.send = adin2111_netif_send_,
+                                              .enable = adin2111_netif_enable_,
+                                              .disable =
+                                                  adin2111_netif_disable_};
```

Comment:
Will the compiler actually inline these functions if they are callbacks?

Reply:
I don't know, but the `inline` specifier is only a hint anyway. It certainly could, but there's no problem if it doesn't.